File: 2022/18.cc
#include "common.h"
#include "dense_set.h"

namespace aoc_2022_18 {

using Cube = std::array<int, 3>;

static Cube with_offset(Cube c, int dim, int delta)
{
    c[dim] += delta;
    return c;
}

static int sum_surface_area(const dense_set<Cube, CrcHasher> &cubes)
{
    int sum = 0;
    for (auto &c : cubes) {
        int free = 6;
        for (int dim = 0; dim < 3; dim++) {
            for (int sign : {-1, 1}) {
                if (cubes.count(with_offset(c, dim, sign)))
                    free--;
            }
        }
        sum += free;
    }

    return sum;
}

static dense_set<Cube, CrcHasher> flood(dense_set<Cube, CrcHasher> occupied)
{
    int min[3] = {INT_MAX, INT_MAX, INT_MAX};
    int max[3] = {INT_MIN, INT_MIN, INT_MIN};

    for (auto &c : occupied) {
        for (int i = 0; i < 3; i++) {
            min[i] = std::min(min[i], c[i]);
            max[i] = std::max(max[i], c[i]);
        }
    }

    auto can_fill = [&](Cube c) {
        // Can't fill out of bounds
        for (int i = 0; i < 3; i++) {
            if (!(min[i] <= c[i] && c[i] <= max[i]))
                return false;
        }

        // Can only fill empty space, not lava cubes or water/steam
        return occupied.count(c) == 0;
    };

    std::vector<Cube> q{{min[0], min[1], min[2]}};
    while (!q.empty()) {
        auto u = q.back();
        q.pop_back();
        occupied.insert(u);

        for (int dim = 0; dim < 3; dim++) {
            for (int sign : {-1, 1}) {
                Cube v = with_offset(u, dim, sign);
                if (can_fill(v))
                    q.push_back(v);
            }
        }
    }

    dense_set<Cube, CrcHasher> unvisited;
    unvisited.reserve(2048);

    for (int x = min[0]; x < max[0]; x++) {
        for (int y = min[1]; y < max[1]; y++) {
            for (int z = min[2]; z < max[2]; z++) {
                Cube p = {x, y, z};
                if (!occupied.count(p))
                    unvisited.emplace(p);
            }
        }
    }

    return unvisited;
}

void run(std::string_view buf)
{
    const auto lines = split_lines(buf);
    dense_set<Cube, CrcHasher> cubes;
    cubes.reserve(lines.size());
    for (std::string_view line : lines)
        cubes.insert(find_numbers_n<int, 3>(line));

    const auto area = sum_surface_area(cubes);
    fmt::print("{}\n", area);

    const auto unvisited = flood(cubes);
    fmt::print("{}\n", area - sum_surface_area(unvisited));
}

}

File: 2022/18.py
#!/usr/bin/env python3
import itertools
import sys


def replace(t, i, v):
    return t[:i] + (v,) + t[i + 1 :]


def sum_surface_area(cubes):
    occlusion = {}

    for c in cubes:
        for dim in (0, 1, 2):
            for db, d in ((0, -1), (1, 1)):
                n = replace(c, dim, c[dim] + d)
                if n in cubes:
                    occlusion[c] = occlusion.get(c, 0) | (1 << (2 * dim + db))

    return sum(6 - occlusion.get(c, 0).bit_count() for c in cubes)


def flood(cubes):
    xmin = min(c[0] for c in cubes) - 1
    xmax = max(c[0] for c in cubes) + 1
    ymin = min(c[1] for c in cubes) - 1
    ymax = max(c[1] for c in cubes) + 1
    zmin = min(c[2] for c in cubes) - 1
    zmax = max(c[2] for c in cubes) + 1

    occupied = set(cubes)

    def can_fill(c):
        # Can't fill out of bounds
        if not (xmin <= c[0] <= xmax and ymin <= c[1] <= ymax and zmin <= c[2] <= zmax):
            return False

        # Can only fill empty space, not lava cubes or water/steam
        return w not in occupied

    q = [(xmin, ymin, zmin)]
    while q:
        v = q.pop()
        occupied.add(v)

        for dim, d in itertools.product((0, 1, 2), (-1, 1)):
            w = replace(v, dim, v[dim] + d)
            if can_fill(w):
                q.append(w)

    all_cubes = itertools.product(
        range(xmin + 1, xmax), range(ymin + 1, ymax), range(zmin + 1, zmax)
    )
    return set(c for c in all_cubes if c not in occupied)


def main():
    lines = sys.stdin.readlines()
    cubes = set(tuple(map(int, line.split(","))) for line in lines)

    area = sum_surface_area(cubes)
    print(area)

    unvisited = flood(cubes)
    print(area - sum_surface_area(unvisited))


if __name__ == "__main__":
    main()
